src/node/jsonrpc.rs

```diff
@@ -0,0 +1,172 @@
+use crate::node::config::{IPCAgentJsonRPCNodeConfig, DEFAULT_JSON_RPC_ENDPOINT};
+use bytes::Bytes;
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+use warp::http::StatusCode;
+use warp::reject::Reject;
+use warp::reply::with_status;
+use warp::{Filter, Rejection, Reply};
+
+/// The json rpc request param. It is the standard form our json-rpc and follows a structure similar
+/// to the one of the Ethereum RPC: https://ethereum.org/en/developers/docs/apis/json-rpc/#curl-examples
+#[derive(Serialize, Deserialize, Debug)]
+struct JSONRPCRequest {
+    pub id: u16,
+    pub jsonrpc: String,
+    pub method: String,
+    pub params: Value,
+}
+
+/// The json rpc response. It is the standard form our json-rpc and follows a structure similar
+/// to the one of the Ethereum RPC: https://ethereum.org/en/developers/docs/apis/json-rpc/#curl-examples
+#[derive(Debug, Serialize, Deserialize)]
+struct JSONRPCResponse<T: Serialize> {
```

Comment:
This must also include an `error` field to conform with the JSON-RPC spec (see https://www.jsonrpc.org/specification#response_object). The serialization then uses one of `result` and `error` depending on the outcome of the processing of the request.

Reply:
right right, then in this case I think we can do something like this:
```
struct JSONRPCResponse<T: Serialize> {
    pub id: u16,
    pub jsonrpc: String,
    #[some serde config to ignore serialization if None]
    pub result: Option<T>,
    #[some serde config to ignore serialization if None]
    pub error: Option<String>, // we can use T is we want support struct.
}
```
Then, we can based to `Result` to set `result` or `error`.

Reply:
Update: 2023/02/20:
I think it's better not to mix the result and error. I read the [json rpc spec](https://www.jsonrpc.org/specification#response_object) of json_rpc, seems the error can be more complex. It's neater to use another struct just to handle the error.

Reply:
You can simply have here a field like: 
```
pub error: Option<JsonRpcError>
```
Is this what you're suggesting?

Reply:
Refer to this [struct](https://github.com/consensus-shipyard/ipc-agent/blob/1187b226e389d5323c8edc3f02f506534f3a683c/src/server/response.rs#L42). The usage is [here](https://github.com/consensus-shipyard/ipc-agent/blob/1187b226e389d5323c8edc3f02f506534f3a683c/src/server/jsonrpc.rs#L76)

Reply:
As discussed, will close this one first then sharing the request and responses. Resolving first.